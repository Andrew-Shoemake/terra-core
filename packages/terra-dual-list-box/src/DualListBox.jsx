import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import IconRight from 'terra-icon/lib/icon/IconRight';
import IconLeft from 'terra-icon/lib/icon/IconLeft';
import classNames from 'classnames/bind';
import ListBox from './_ListBox';
import DualListUtils from './_DualListBoxUtils';
import styles from './DualListBox.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Child Nodes that are rendered as target listbox.
   */
  source: PropTypes.node,
  /**
   * Child Nodes that are rendered as target listbox.
   */
  target: PropTypes.node,
};

const defaultProps = {
  source: undefined,
  target: undefined,
};

const DualListBox = (props) => {
  const {
    source,
    target,
  } = props;

  const [sourceList, setSourceList] = useState(source);
  const [targetList, setTargetList] = useState(target);

  useEffect(() => {
    DualListUtils.clearSelections();
  }, [sourceList, targetList]);

  const iconLeftOnClick = () => {
    const { targetItems, sourceItems } = (DualListUtils.moveItems(targetList, sourceList));
    setSourceList(targetItems);
    setTargetList(sourceItems);
  };

  const iconRightOnClick = () => {
    const { targetItems, sourceItems } = (DualListUtils.moveItems(sourceList, targetList));
    setTargetList(targetItems);
    setSourceList(sourceItems);
  };

  return (
    <div className={cx('container')}>
      <ListBox headerLabelText="Source">
        {sourceList}
      </ListBox>
      <div className={cx('button-wrapper')}>
        <button className={cx('arrow-button')} type="button" onClick={iconLeftOnClick}>
          <IconLeft />
        </button>
        <button className={cx('arrow-button')} type="button" onClick={iconRightOnClick}>
          <IconRight />
        </button>
      </div>
      <ListBox headerLabelText="Target">
        {targetList}
      </ListBox>
    </div>
  );
};

DualListBox.propTypes = propTypes;
DualListBox.defaultProps = defaultProps;
export default DualListBox;
