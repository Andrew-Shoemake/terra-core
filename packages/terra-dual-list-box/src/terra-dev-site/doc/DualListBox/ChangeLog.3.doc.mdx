import { Badge } from 'terra-dual-list-box/package.json?dev-site-package';

import ChangeLog from 'terra-dual-list-box/CHANGELOG.md';

<Badge />

<ChangeLog />
