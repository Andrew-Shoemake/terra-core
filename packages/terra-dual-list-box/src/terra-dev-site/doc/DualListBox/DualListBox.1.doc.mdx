import { Badge } from 'terra-dual-list-box/package.json?dev-site-package';

import DefaultExample from '../example/DefaultDualListBox?dev-site-example';

import DualListBoxPropsTable from 'terra-dual-list-box/src/DualListBox?dev-site-props-table';
import ListItemPropsTable from 'terra-dual-list-box/src/ListItem?dev-site-props-table';

<Badge />

# Terra Dual List Box

Terra-Dual-List-Box enables user to transfer the elements between two list box in an intuitive and efficient way.

## Getting Started

- Install with [npmjs][3]:
  - `npm install terra-dual-list-box`

<!-- AUTO-GENERATED-CONTENT:START Peer Dependencies -->
## Peer Dependencies

This component requires the following peer dependencies be installed in your app for the component to properly function.

| Peer Dependency | Version |
|-|-|
| react | ^16.8.5 |
| react-dom | ^16.8.5 |
| react-intl | ^2.8.0 |



<!-- AUTO-GENERATED-CONTENT:END -->

## Implementation Notes:
The Dual list box component must be composed inside the [Base][1] component with a locale in order for it to load the correct translation strings.

[1]: https://github.com/cerner/terra-core/blob/master/packages/terra-base/src/terra-dev-site/doc/base/Base.1.doc.mdx
[3]: https://www.npmjs.com

## Component Features
* [Cross-Browser Support](https://engineering.cerner.com/terra-ui/contributing/terra-ui/component-standards#cross-browser-support)
* [Responsive Support](https://engineering.cerner.com/terra-ui/contributing/terra-ui/component-standards#responsive-support)
* [Mobile Support](https://engineering.cerner.com/terra-ui/contributing/terra-ui/component-standards#mobile-support)
* [Internationalization Support](https://engineering.cerner.com/terra-ui/contributing/terra-ui/component-standards#internationalization-i18n-support)
* [Localization Support](https://engineering.cerner.com/terra-ui/contributing/terra-ui/component-standards#internationalization-i18n-support)

[View component source code](https://github.com/cerner/terra-core/tree/master/packages/terra-dual-list-box)

## Examples
<DefaultExample title="Default Dual List Box" />

## Dual List Box Props
<DualListBoxPropsTable />

## List Item Props
<ListItemPropsTable />
