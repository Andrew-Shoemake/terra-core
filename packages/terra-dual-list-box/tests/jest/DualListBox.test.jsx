/* eslint-disable no-console */
import React from 'react';
import DualListBox from '../../src/DualListBox';

describe('DualListBox with no props', () => {
  // Snapshot Tests
  it('should render a default component', () => {
    const wrapper = <DualListBox />;
    expect(wrapper).toMatchSnapshot();
  });
});
