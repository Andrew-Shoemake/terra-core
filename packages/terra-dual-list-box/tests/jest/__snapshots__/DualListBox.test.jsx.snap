// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DualListBox with no props should render a default component 1`] = `<DualListBox />`;
